hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerPrepareState.java

```diff
@@ -0,0 +1,279 @@
+   */
+  public synchronized void finishPrepare(long index) throws IOException {
+    finishPrepare(index, true);
+  }
+
+  private void finishPrepare(long index, boolean writeFile) throws IOException {
+    // Enabling the prepare gate is idempotent, and may have already been
+    // performed if we are the leader.If we are a follower, we must ensure this
+    // is run now case we become the leader.
+    enablePrepareGate();
+
+    if (writeFile) {
+      writePrepareMarkerFile(index);
+    }
+    prepareIndex = index;
+    status = PrepareStatus.PREPARE_COMPLETED;
+  }
+
+  /**
+   * Uses the on disk marker file to determine the OM's prepare state.
+   * If the marker file exists and contains an index matching {@code
+   * expectedPrepareIndex}, the necessary steps will be taken to finish
+   * preparation and the state will be moved to
+   * {@link PrepareStatus#PREPARE_COMPLETED}.
+   * Else, the status will be moved to
+   * {@link PrepareStatus#PREPARE_NOT_STARTED} and any preparation steps will
+   * be cancelled.
+   *
+   * @return The status the OM is in after this method call.
+   * @throws IOException If the marker file cannot be read, and it cannot be
+   * deleted as part of moving to the
+   * {@link PrepareStatus#PREPARE_NOT_STARTED} state.
+   */
+  public synchronized PrepareStatus restorePrepare(long expectedPrepareIndex)
+      throws IOException {
+    boolean prepareIndexRead = true;
+    long prepareMarkerIndex = NO_PREPARE_INDEX;
+
+    File prepareMarkerFile = getPrepareMarkerFile();
+    if (prepareMarkerFile.exists()) {
+      byte[] data = new byte[(int) prepareMarkerFile.length()];
+      try(FileInputStream stream = new FileInputStream(prepareMarkerFile)) {
+        stream.read(data);
+      } catch (IOException e) {
+        LOG.error("Failed to read prepare marker file {} while restoring OM.",
+            prepareMarkerFile.getAbsolutePath());
+        prepareIndexRead = false;
+      }
+
+      try {
+        prepareMarkerIndex = Long.parseLong(
+            new String(data, StandardCharsets.UTF_8));
+      } catch (NumberFormatException e) {
+        LOG.error("Failed to parse log index from prepare marker file {} " +
+            "while restoring OM.", prepareMarkerFile.getAbsolutePath());
+        prepareIndexRead = false;
+      }
+    } else {
+      // No marker file found.
+      prepareIndexRead = false;
+    }
+
+    boolean prepareRestored = false;
+    if (prepareIndexRead) {
+      if (prepareMarkerIndex != expectedPrepareIndex) {
+        LOG.error("Failed to restore OM prepare state, because the expected " +
+            "prepare index {} does not match the index {} written to the " +
+            "marker file.", expectedPrepareIndex, prepareMarkerIndex);
+      } else {
+        // Prepare state can only be restored if we read the expected index
+        // from the marker file.
+        prepareRestored = true;
+      }
+    }
+
+    if (prepareRestored) {
+      enablePrepareGate();
```

Comment:
There is an 'enablePrepareGate()' in finishPrepare. Do we need the one in Line # 180?